pkg/operator/ceph/cluster/newosd/osd.go

```diff
@@ -0,0 +1,159 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package newosd
+
+import (
+	"fmt"
+
+	"github.com/coreos/pkg/capnslog"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	rookalpha "github.com/rook/rook/pkg/apis/rook.io/v1alpha2"
+	"github.com/rook/rook/pkg/clusterd"
+	cephconfig "github.com/rook/rook/pkg/daemon/ceph/config"
+	opspec "github.com/rook/rook/pkg/operator/ceph/spec"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "op-osd")
+
+const (
+	appName = "rook-ceph-osd"
+
+	prepareAppName    = "rook-ceph-osd-prepare"
+	prepareAppNameFmt = "rook-ceph-osd-prepare-%s"
+
+	osdAppNameFmt = "rook-ceph-osd-%d"
+
+	osdLabelKey                  = "ceph-osd-id"
+	clusterAvailableSpaceReserve = 0.05
+	serviceAccountName           = "rook-ceph-osd"
+	unknownID                    = -1
+
+	cephOsdPodMinimumMemory uint64 = 4096 // minimum amount of memory in MB to run the pod
+)
+
+// Controller keeps track of the OSDs
+type Controller struct {
+	context     *clusterd.Context
+	namespace   string
+	cephCluster *cephCluster
+	rookImage   string
+	ownerRef    metav1.OwnerReference
+}
+
+type cephCluster struct {
+	spec *cephv1.ClusterSpec
+	info *cephconfig.ClusterInfo
+}
+
+func NewController(
+	context *clusterd.Context,
+	namespace string,
+	cephSpec *cephv1.ClusterSpec,
+	cephInfo *cephconfig.ClusterInfo,
+	ownerRef metav1.OwnerReference,
+) *Controller {
+	return &Controller{
+		context:   context,
+		namespace: namespace,
+		cephCluster: &cephCluster{
+			spec: cephSpec,
+			info: cephInfo,
+		},
+		ownerRef: ownerRef,
+	}
+}
+
+// Start the osd management
+func (c *Controller) Start() error {
+	logger.Infof("OSD controller started for namespace %s", c.namespace)
+
+	// Validate pod's memory if specified
+	r := cephv1.GetOSDResources(c.cephCluster.spec.Resources)
+	err := opspec.CheckPodMemory(r, cephOsdPodMinimumMemory)
+	if err != nil {
+		return fmt.Errorf("user-requested OSD pod resource requirements are insufficient. %+v", err)
+	}
+
+	provisionableNodes, err := c.getProvisionableNodes()
+	if err != nil {
+		return fmt.Errorf("OSD controller failed. %+v", err)
+	}
+	logger.Debugf("returned provisionable nodes: %+v", provisionableNodes)
+
+	// TODO: inventory provisionable nodes
+
+	// TODO: provision new OSDs on provisionable nodes with provisionable disks
+
+	// TODO: find nodes already provisioned with OSDs ???
+
+	// TODO: get osd inventory from nodes that are newly provisionable and nodes which are already
+	// provisioned with osds
+
+	// TODO: start OSDs on the OSD inventory from above
+
+	// TODO: what does automatic osd removal look like now?
```

Comment:
No more automatic osd removal, right?

Reply:
Well ... automatic OSD removal I think could mean many things. I believe we all agree that we are not removing the OSDs from CRUSH. But we are (or may be) removing pods, which is more what I meant here. Bigger picture, I guess this means, what does it look like when an osd is marked `safe-to-remove` in this scheme? The end goal is that we keep a record of which nodes have had OSDs deployed on them. My current thougts are 2 options: scale deployments down to 0 (Rook will use deployments to have that record of nodes+osds) or to inspect Ceph's CRUSH / OSD map(s) to find nodes on which OSDs may previously have existed (this option is probably better if it's not overly complicated).

As a future consideration, I do see some benefit to automatically removing OSDs completely (from CRUSH and otherwise) as long as users enable some option in the CRD and Rook allows for a very long timeout -- I've been arbitrarily suggesting 45 days. I think that might still be of use to novice users.